### ISSUE TITLE
CLI - Financial Statistics

### ISSUE THREAD

**Issue by dnwiebe** [MASQ-Project/MASQ-Node-issues#236]
Enhance the CLI and the Node to be able to handle the 
	GetFinancialStatistics/FinancialStatisticsResponse messages, or their 
	replacements.

Reference GH-202.

This card will encompass #493

The GUI team needs UI messaging so we can know the following financial information from the accountant:

- [x] #590
- [x] MASQ Incoming confirmed total
- [x] MASQ Outgoing pending
- [x] MASQ Outgoing confirmed totals

This relates to GUI issue:
https://github.com/MASQ-Project/MASQ-React-App/issues/91

**Comment by anonmate** [comment]
In the UI we have:

• total pending income
• total pending debt
• total earned in last 1hr
• total spent in last 1hr
• total earned all time
• total spent all time

We can create most of these from just having websockets messaging of payments as they go in/out with values – and we can store the totals if you really don't want to in the DB. As well as do the maths on how much spent/earned in X amount of time.

We just need pendings and successful transactions for spent and earned, I believe we can do the rest

It might also be good to know what the user was paying for in his transaction (how much data or something related to ratepack statistics). A large problem we face is not being able to show the user the price of the data he consumed.

If it is fixed in ratepack this problem largely disappears. But when/if we have a free market and a price oracle, this would need to be addressed

**Comment by anonmate** [comment]
@bertllll @kauri-hero @dnwiebe 

<img width="1552" alt="Screenshot 2022-01-25 at 23 31 15" src="https://user-images.githubusercontent.com/77641959/151077318-e53cf679-af1c-4efb-99f0-7550c6655385.png">
<img width="1123" alt="Screenshot 2022-01-25 at 23 32 25" src="https://user-images.githubusercontent.com/77641959/151077364-237aa590-4f73-40cc-a786-f5a2d2b055d6.png">

**Comment by dnwiebe** [comment]
Suggestion--make sure @anonmate reads this and agrees:

The Node already keeps track of these numbers:
* Payable accounts over a minimum balance and younger than a minimum age
* Total payable amount (including accounts that don't pass the above test)
* Receivable accounts over a minimum balance and younger than a minimum age
* Total receivable amount (including accounts that don't pass the above test)

It should be extended to also keep track of more numbers. Consider the following events:
* The Proxy Server reports to the Accountant that it has ordered exit services from a Node
* The Proxy Server reports to the Accountant that it has ordered routing services from a Node
* The Proxy Client reports to the Accountant that it has provided exit services to a Node
* The Hopper reports to the Accountant that it has provided routing services to a Node

For each of these events, the Accountant should keep running totals of the following figures, starting with zeros at startup:
* Number of CORES packages
* Number of payload bytes
* Payable or receivable amounts for CORES packages
* Payable or receivable amounts for bytes

Deriving the first four ("already keeps track of") numbers requires a set of database scans and is reasonably resource-intensive. However, once the outlined changes are made to the Accountant, it will have the final sixteen numbers always ready in memory, so it should be quick and easy to provide them.

Hence we should end up with two messages, one for the four numbers calculated since installation, and another for the sixteen numbers calculated since startup.

**Comment by anonmate** [comment]
@dnwiebe As far as what the UI needs, it's purely:

- Total payable amount (We call this pending debts in the UI)
- Total receivable amount (We call this pending earnings in the UI)
- Total paid amount (We call this total spent in the UI) 
- Total received amount (We call this total earned in the UI)

Though in fact from a user perspective, pending earnings and debts are most probably confusing as they think 'why aren't they just settled, why is the UI even telling me this?'. So it's possible we eventually don't have these either, or at least, they become more hidden.

**Comment by dnwiebe** [comment]
Okay, @anonmate and @bertllll , how about this?

We give you the four numbers you're asking for, except that the last two numbers are totals _since the Node started up,_ and they get zeroed out every time the Node is stopped and started again.

Then, in a separate card--for which Bert has already written some code--we add a couple of places in the database where all-time totals can be stored, and we put in a facility whereby whenever the Node panics, those places in the database are automatically updated with the since-startup totals before the panic is allowed to complete.

Would that be satisfactory?

**Comment by bertllll** [comment]
Please @anonmate answer these questions for me.

Until now UiFinancials is a sketch that has never gone alive, well, not fully, we've actually got some pieces of code. Now the question is how to translate the existing code into your vision which is, as to my understanding, partly diverging.

#### `financials`

##### Direction: Request
##### Correspondent: Node
##### Layout:
```
"payload": {
    "payableMinimumAmount" = <nonnegative integer>,
    "payableMaximumAge" = <nonnegative integer>,
    "receivableMinimumAmount" = <nonnegative integer>,
    "receivableMaximumAge" = <nonnegative integer>
}
```
##### Direction: Response
##### Correspondent: Node
##### Layout:
```
"payload": {
    "payables": [
        {
            "wallet": <string>,
            "age": <nonnegative integer>,
            "amount": <nonnegative integer>,
            "pendingPayableHashOpt": <optional string>
        },
        < ... >
    ],
    "totalPayable": <nonnegative integer>,
    "receivables": [
        {
            "wallet": <string>,
            "age": <nonnegative integer>,
            "amount": <nonnegative integer>
        },
        < ... >
    ],
    "totalReceivable": <nonnegative integer>
}
```
Whoever designed it, he was likely to plan to allow fetching detailed records of individual payable and receivable records. I have no idea if it could be useful, but I want to hear from you if I should go ahead and throw this code away. I'm supposing you've got a strategy for this kind of things and so you might be able to look further into the future. 

Do you think I should remake this messages to have its request version being empty (as we may not need to parametrize anything there under these conditions) and a response version with those four values of various totals which we talked over before?

Also, this question is perhaps groundless, but do you plan to have a single pair of messages associated with financials or do you think it will be divided into more as you demand to know more information going forward?

**Comment by anonmate** [comment]
> Okay, @anonmate and @bertllll , how about this?
> 
> We give you the four numbers you're asking for, except that the last two numbers are totals _since the Node started up,_ and they get zeroed out every time the Node is stopped and started again.
> 
> Then, in a separate card--for which Bert has already written some code--we add a couple of places in the database where all-time totals can be stored, and we put in a facility whereby whenever the Node panics, those places in the database are automatically updated with the since-startup totals before the panic is allowed to complete.
> 
> Would that be satisfactory?

Yes, absolutely

**Comment by anonmate** [comment]
> Also, this question is perhaps groundless, but do you plan to have a single pair of messages associated with financials or do you think it will be divided into more as you demand to know more information going forward?

Honestly, I'm just a visual designer and deciding on such important factors of code architecture doesn't feel right for me to do. But if I'm to answer your questions in some sense, then it would be this: one message that returns a JSON object with all financial information should be fine. I don't see a need for multiple messages when everything can be retrieved with one - unless it ends up becoming much larger in the future.

Basically as long as we have the data in the 4 fields I've asked for, accessible inside a JSON object, we'll be fine

**Comment by kauri-hero** [comment]
Agree absolutely with your request for a JSON object, since it can be structured to hold the aspects of financial data already discussed (currently those 4 items since the Node has started) KauriHeroProject OwnerMASQ NetworkFrom: anonmate ***@***.***>Sent: Thursday, 24 February 2022, 11:29 pmTo: MASQ-Project/Node-issuesCc: KauriHero; MentionSubject: Re: [MASQ-Project/Node-issues] CLI - Financial Statistics (#236)

Also, this question is perhaps groundless, but do you plan to have a single pair of messages associated with financials or do you think it will be divided into more as you demand to know more information going forward?

Honestly, I'm just a visual designer and deciding on such important factors of code architecture doesn't feel right for me to do. But if I'm to answer your questions in some sense, then it would be this: one message that returns a JSON object with all financial information should be fine. I don't see a need for multiple messages when everything can be retrieved with one - unless it ends up becoming much larger in the future.
Basically as long as we have the data in the 4 fields I've asked for, accessible inside a JSON object, we'll be fine

**Comment by bertllll** [comment]
**Test plan**

The functionality give by this card isn't broad, it's a start of another development area. However we can start with just limited quality. Other cards will follow (not prioritized now) later to pick up on this elementary one.
During my work on this I realized we need more than a simple counter of growing totals, it wouldn't work perfectly because of certain corner cases. This will involve advanced measures put in the code to make sure we're able to bridge times when the Node is going off and, forgetting all data from the runtime memory. We also have to handle somehow those unexpected fatal events that may be caused by panics. This is not trivial and so that's why it blew up the time spent on this first-sight seemingly quick card. (As response to this two new big cards have been created). 

I'm sorry that the functionality is now dump but it definitely has a potential of grows. 

**IMPORTANT DISCLAIMER OUT OF THE STATED ABOVE:**
Although we talk about all time totals in case of "already paid" money for receivable or payable, we really cannot aspire to count it like so, always since the birth, right now. For that we would need those other cards completed. So there's been a tactical decision to keep talking about it as if it were all-time value but for a limited time it will only refer to a value counted since the last startup. After the Nodes goes down and up again this value starts again at a zero. This also mean these values are not stored in the database for now, which will change later though. 

Possible items to be checked in the test:

You just need to start up your Node and while it's running you can check the financial statuses using `financials` which is the new command. 
You should then see a set of four values. 
Try to consume data and hopefully also allow conditions for paying your debts. Both should be taken track of within the figures returned by the command's response; you should be able to see that this happens for the side of payables as well as for receivables (It just depends on activity of your Node).

**ANOTHER DISCLAIMER:** I remarked myself a big latency before you get the response, if too long, you will time out and no response ever comes back (which is really bad). So it's possible that sometimes you will order the command but it will just say your ordered procedure timed out. One other card should solve this in the future (where we will enforce Accountant in his job by adding another Actor collecting continuously information that might be at some point requested from the UI). Just make sure that most of time you can get the response coming back.
